`ArrayIndexOutOfBoundsException` during core graph preparation for hgv

As [reported on the ask forum](https://ask.openrouteservice.org/t/java-util-concurrent-executionexception-java-lang-arrayindexoutofboundsexception-index-1048576-out-of-bounds-for-length-1048576/3932), build of [bengaluru.pbf](https://easyupload.io/fo7nul) against [app.config.txt](https://github.com/GIScience/openrouteservice/files/8833360/app.config.txt) fails with the following error.

The underlying problem might be the same as for a [similar issue](https://github.com/GIScience/openrouteservice/issues/1149) reported earlier, which unfortunately we couldn't reproduce.

```
03 Jun 17:13:41 WARN [extensions.ORSOSMReader] - Index 1048576 out of bounds for length 1048576. Way id = 363083979
03 Jun 17:14:06 ERROR [routing.RoutingProfileManager] - java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.ArrayIndexOutOfBoundsException: Index 1048576 out of bounds for length 1048576
03 Jun 17:14:06 ERROR [routing.RoutingProfileManager] - Failed to initialize RoutingProfileManager instance.
java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.ArrayIndexOutOfBoundsException: Index 1048576 out of bounds for length 1048576
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.heigit.ors.routing.RoutingProfileManager.initialize(RoutingProfileManager.java:178) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfileManager.getInstance(RoutingProfileManager.java:72) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfileManager.isInitComplete(RoutingProfileManager.java:668) ~[classes/:?]
	at org.heigit.ors.servlet.listeners.ORSKafkaConsumerInitContextListener.lambda$contextInitialized$0(ORSKafkaConsumerInitContextListener.java:47) ~[classes/:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.ArrayIndexOutOfBoundsException: Index 1048576 out of bounds for length 1048576
	at org.heigit.ors.routing.graphhopper.extensions.core.CoreAlgoFactoryDecorator.prepare(CoreAlgoFactoryDecorator.java:314) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.prepareCore(ORSGraphHopper.java:787) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.postProcessing(ORSGraphHopper.java:690) ~[classes/:?]
	at com.graphhopper.GraphHopper.process(GraphHopper.java:703) ~[graphhopper-core-v0.13.21.jar:?]
	at com.graphhopper.GraphHopper.importOrLoad(GraphHopper.java:680) ~[graphhopper-core-v0.13.21.jar:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.importOrLoad(ORSGraphHopper.java:169) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfile.initGraphHopper(RoutingProfile.java:182) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfile.<init>(RoutingProfile.java:126) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfileLoader.call(RoutingProfileLoader.java:35) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfileLoader.call(RoutingProfileLoader.java:21) ~[classes/:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	... 1 more
Caused by: java.util.concurrent.ExecutionException: java.lang.ArrayIndexOutOfBoundsException: Index 1048576 out of bounds for length 1048576
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.heigit.ors.routing.graphhopper.extensions.core.CoreAlgoFactoryDecorator.prepare(CoreAlgoFactoryDecorator.java:310) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.prepareCore(ORSGraphHopper.java:787) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.postProcessing(ORSGraphHopper.java:690) ~[classes/:?]
	at com.graphhopper.GraphHopper.process(GraphHopper.java:703) ~[graphhopper-core-v0.13.21.jar:?]
	at com.graphhopper.GraphHopper.importOrLoad(GraphHopper.java:680) ~[graphhopper-core-v0.13.21.jar:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.importOrLoad(ORSGraphHopper.java:169) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfile.initGraphHopper(RoutingProfile.java:182) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfile.<init>(RoutingProfile.java:126) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfileLoader.call(RoutingProfileLoader.java:35) ~[classes/:?]
	at org.heigit.ors.routing.RoutingProfileLoader.call(RoutingProfileLoader.java:21) ~[classes/:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	... 1 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1048576 out of bounds for length 1048576
	at com.graphhopper.util.BitUtilLittle.toShort(BitUtilLittle.java:32) ~[graphhopper-core-v0.13.21.jar:?]
	at com.graphhopper.storage.RAMDataAccess.getShort(RAMDataAccess.java:225) ~[graphhopper-core-v0.13.21.jar:?]
	at org.heigit.ors.routing.graphhopper.extensions.storages.HeavyVehicleAttributesGraphStorage.hasEdgeRestriction(HeavyVehicleAttributesGraphStorage.java:163) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.edgefilters.core.HeavyVehicleCoreEdgeFilter.accept(HeavyVehicleCoreEdgeFilter.java:33) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.edgefilters.EdgeFilterSequence.accept(EdgeFilterSequence.java:29) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.core.PrepareCore.prepareNodes(PrepareCore.java:190) ~[classes/:?]
	at org.heigit.ors.routing.graphhopper.extensions.core.PrepareCore.doSpecificWork(PrepareCore.java:175) ~[classes/:?]
	at com.graphhopper.routing.util.AbstractAlgoPreparation.doWork(AbstractAlgoPreparation.java:30) ~[graphhopper-core-v0.13.21.jar:?]
	at org.heigit.ors.routing.graphhopper.extensions.core.CoreAlgoFactoryDecorator.lambda$prepare$0(CoreAlgoFactoryDecorator.java:300) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	... 1 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -47,6 +47,7 @@ RELEASING:
 - added missing matchTraffic override ([#1133](https://github.com/GIScience/openrouteservice/issues/1133))
 - typo in docker documentation
 - foot routing via `waterway=lock_gate` ([#1177](https://github.com/GIScience/openrouteservice/issues/1177))
+- address data alignment issue in hgv extended storage which occasionally caused `ArrayIndexOutOfBoundsException` ([#1181](https://github.com/GIScience/openrouteservice/issues/1181))
 
 ## [6.7.0] - 2022-01-04
 ### Added

--- a/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/storages/HeavyVehicleAttributesGraphStorage.java
+++ b/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/storages/HeavyVehicleAttributesGraphStorage.java
@@ -17,6 +17,7 @@ import com.graphhopper.storage.DataAccess;
 import com.graphhopper.storage.Directory;
 import com.graphhopper.storage.Graph;
 import com.graphhopper.storage.GraphExtension;
+import com.graphhopper.util.BitUtil;
 import org.heigit.ors.routing.graphhopper.extensions.VehicleDimensionRestrictions;
 
 public class HeavyVehicleAttributesGraphStorage implements GraphExtension {
@@ -32,6 +33,8 @@ public class HeavyVehicleAttributesGraphStorage implements GraphExtension {
 	protected int edgesCount;
 
 	private static final double FACTOR = 100.0;
+
+	private final byte[] buffer = new byte[2];
 
 	public HeavyVehicleAttributesGraphStorage(boolean includeRestrictions) {
 		efVehicleType = nextBlockEntryIndex(1);
@@ -113,7 +116,8 @@ public class HeavyVehicleAttributesGraphStorage implements GraphExtension {
 
 		for (int i = 0; i < VehicleDimensionRestrictions.COUNT; i++) {
 			short shortValue = (short) (restrictionValues[i] * FACTOR);
-			orsEdges.setShort(edgePointer + efRestrictions + i * EF_RESTRICTION_BYTES, shortValue);
+			BitUtil.LITTLE.fromShort(buffer, shortValue);
+			orsEdges.setBytes(edgePointer + efRestrictions + i * EF_RESTRICTION_BYTES, buffer, 2);
 		}
 	}
 
@@ -123,7 +127,7 @@ public class HeavyVehicleAttributesGraphStorage implements GraphExtension {
 		if (efRestrictions == -1)
 			throw new IllegalStateException(MSG_EF_RESTRICTION_IS_NOT_SUPPORTED);
 
-		return orsEdges.getShort(edgeBase + efRestrictions + valueIndex * EF_RESTRICTION_BYTES) / FACTOR;
+		return getShort(edgeBase + efRestrictions + valueIndex * EF_RESTRICTION_BYTES) / FACTOR;
 	}
 
 	public boolean getEdgeRestrictionValues(int edgeId, double[] retValues) {
@@ -133,9 +137,14 @@ public class HeavyVehicleAttributesGraphStorage implements GraphExtension {
 			throw new IllegalStateException(MSG_EF_RESTRICTION_IS_NOT_SUPPORTED);
 
 		for (int i = 0; i < VehicleDimensionRestrictions.COUNT; i++)
-			retValues[i] = orsEdges.getShort(edgeBase + efRestrictions + i * EF_RESTRICTION_BYTES) / FACTOR;
+			retValues[i] = getShort(edgeBase + efRestrictions + i * EF_RESTRICTION_BYTES) / FACTOR;
 
 		return true;
+	}
+
+	private short getShort(long bytePos) {
+		orsEdges.getBytes(bytePos, buffer, 2);
+		return BitUtil.LITTLE.toShort(buffer);
 	}
 
 	public int getEdgeVehicleType(int edgeId, byte[] buffer) {
@@ -160,7 +169,7 @@ public class HeavyVehicleAttributesGraphStorage implements GraphExtension {
 
 		if (efRestrictions > 0)
 			for (int i = 0; i < VehicleDimensionRestrictions.COUNT; i++)
-				if (orsEdges.getShort(edgeBase + efRestrictions + i * EF_RESTRICTION_BYTES) != 0)
+				if (getShort(edgeBase + efRestrictions + i * EF_RESTRICTION_BYTES) != 0)
 					return true;
 
 		return false;

